avocado/plugins/assets.py

```diff
@@ -0,0 +1,179 @@
+                    self.calls.append(call)
+    # visit_call()
+
+    def fetcher(self):
+        """
+        Build an intrumented avocado test with all fetch_asset calls and
+        execute it.
+        """
+        # build raw code from list of calls, if available
+        if len(self.calls) > 0:
+            rcalls = ""
+            for call in self.calls:
+                rcalls = rcalls + call
+        else:
+            LOG_UI.debug('No supported assets found to fetch from %s'
+                         % self.file_name)
+            return
+
+        # temporary test source file
+        fetch_test = tempfile.NamedTemporaryFile(suffix=".py")
+        fetch_test.write(TEST_TEMPLATE.format(fetch_assets=rcalls).encode())
+        fetch_test.flush()
+
+        # run teporary test
+        LOG_UI.debug('Fetching assets from %s' % self.file_name)
+        self.result = process.run("avocado run --json - %s" % fetch_test.name,
+                                  ignore_status=True)
+        fetch_test.close()
+
+        # check if test ran without problems
+        fetch_test_stdout = json.loads(self.result.stdout)
+        if (self.result.exit_status != 0
+                or fetch_test_stdout['pass'] != fetch_test_stdout['total']):
+            LOG_UI.error('%s Check %s for more details.' % (
+                fetch_test_stdout['tests'][0]['fail_reason'],
+                fetch_test_stdout['tests'][0]['logfile']))
+    # fetcher()
+# FetchAssetHandler()
+
+
+class Assets(CLICmd):
+    """
+    Implements the avocado 'assets' subcommand
+    """
+    name = 'assets'
+    description = 'Manage assets'
+
+    def configure(self, parser):
+        """
+        Add the subparser for the assets action.
+
+        :param parser: Main test runner parser.
+        """
+        parser = super(Assets, self).configure(parser)
+
+        subcommands = parser.add_subparsers(dest='assets_subcommand')
+        subcommands.required = True
+
+        fetch_subcommand_parser = subcommands.add_parser('fetch',
+                                help='Fetch assets from test source'\
+                                ' or config file if it\'s not already in the'\
+                                ' cache')
+        fetch_subcommand_parser.add_argument('references', type=str,
+                                             default=[], nargs='*',
+                                             metavar='TEST_REFERENCE',
+                                             help='List of test references')
+    # configure()
+
+    def run(self, config):
+        subcommand = config.get('assets_subcommand')
+
+        if subcommand == 'fetch':
+
+            # check if we have something to work on
+            if len(config['references']) < 1:
+                LOG_UI.error('No test references provided nor any other' \
+                               ' arguments.')
+
+            # fetch assets from instrumented tests
+            fetch_asset_handler = {}
```

Comment:
Hmm, why are you storing the FetchAssetHandlers? I don't see any use of it...

Reply:
The first implementation I had in mind was considering to handle the results from the download out of the object, but in the end, I just implemented it into the same class, so, this indeed is not necessary.